packages/provider-elasticsearch/src/example-types/results/highlighting/response.js

```diff
@@ -0,0 +1,189 @@
+  schema,
+  hit,
+  tags,
+  nestedArrayIncludes = {}
+) => {
+  let arrayOfObjectsPaths = _.keys(getArrayOfObjectsPathsMap(schema))
+
+  let getIndexedArrayObject = (arr, fragments, arrayPath, itemPath) => {
+    let fragmentsMap = _.groupBy(stripTags(tags), fragments)
+    return F.reduceIndexed(
+      (acc, item, index) => {
+        let fragments = fragmentsMap[itemPath ? _.get(itemPath, item) : item]
+        if (fragments) {
+          F.setOn(
+            F.dotJoin([`${index}`, itemPath]),
+            mergeHighlights(tags, ...fragments),
+            acc
+          )
+          for (let itemPath of nestedArrayIncludes[arrayPath] ?? []) {
+            F.updateOn(
+              F.dotJoin([`${index}`, itemPath]),
+              (highlight) => highlight ?? _.get(itemPath, item),
+              acc
+            )
+          }
+        }
+        return acc
+      },
+      arr,
+      _.get(arrayPath, hit._source)
+    )
+  }
+
+  let getArrayPath = (path) =>
+    isArrayOfScalarsField(schema.fields[path])
+      ? path
+      : findByPrefix(path, arrayOfObjectsPaths)
+
+  let highlight = _.flow(
+    groupByMultiField(schema),
+    _.mapValues(_.flatten),
+    F.reduceIndexed((acc, fragments, path) => {
+      let arrayPath = getArrayPath(path)
+      if (arrayPath) {
+        acc[arrayPath] = getIndexedArrayObject(
+          acc[arrayPath] ?? {},
+          fragments,
+          arrayPath,
+          path.slice(arrayPath.length + 1)
+        )
+      } else if (isBlobField(schema.fields[path])) {
+        acc[path] = fragments
+      } else {
+        acc[path] = mergeHighlights(tags, ...fragments)
+      }
+      return acc
+    }, {})
+  )(hit.highlight)
+
+  if (!_.isEmpty(highlight)) hit.highlight = highlight
+}
+
+/**
+ * Remove each path in `paths` from `hit._source`.
+ */
+export let removePathsFromSource = (schema, hit, paths) => {
+  // Nothing to do
+  if (_.isEmpty(paths)) return
+
+  // "aoo" stands for "array of objects", because I was tired of typing it out
+  // over and over again.
+  let aooMap = getArrayOfObjectsPathsMap(schema)
+  let allAooPaths = _.keys(aooMap)
+  let getAooPath = (path) => findByPrefix(path, allAooPaths)
+  let [aooPaths, otherPaths] = _.partition(getAooPath, paths)
+
+  let toRemove = {
+    ...F.arrayToObject(_.identity, _.constant(true), otherPaths),
+    ...F.mapValuesIndexed((paths, aooPath) => {
+      let removeEntireArray =
+        // All nested fields in array of objects should be removed
+        isArraysEqual(paths, aooMap[aooPath]) ||
+        // Or... the path for the array of objects field should be removed
+        _.includes(aooPath, paths)
+      return (
+        removeEntireArray ||
+        _.map((path) => path.slice(aooPath.length + 1), paths)
+      )
+    }, _.groupBy(getAooPath, aooPaths)),
+  }
+
+  let removePathsFromArray = (paths) => (arr) =>
+    _.reduce(
+      (acc, item) => {
+        for (let path of paths) F.unsetOn(path, item)
```

Comment:
> [F.popProperty](https://smartprocure.github.io/futil-js/#popProperty) will return the removed item and might clean this up.

I don't need to return the popped value

> In fact, I think much of this part of the code might just be unsetOnTree

It could, but this fn is more complicated for performance reasons and `unsetOnTree` looks pretty expensive. I'll do post-merge performance test using `unsetOnTree`.

Reply:
This fn is simpler now.